pkg/querier/queryrange/roundtrip.go

```diff
@@ -72,59 +81,73 @@ func MergeMiddlewares(middleware ...Middleware) Middleware {
 	})
 }
 
+// NewTripperware returns a Tripperware configured with middlewares to limit, align, split, retry and cache requests.
+func NewTripperware(cfg Config, log log.Logger, limits Limits, codec Codec, cacheExtractor Extractor) (frontend.Tripperware, error) {
+	queryRangeMiddleware := []Middleware{LimitsMiddleware(limits)}
+	if cfg.AlignQueriesWithStep {
+		queryRangeMiddleware = append(queryRangeMiddleware, InstrumentMiddleware("step_align"), StepAlignMiddleware)
+	}
+	if cfg.SplitQueriesByDay {
+		queryRangeMiddleware = append(queryRangeMiddleware, InstrumentMiddleware("split_by_day"), SplitByDayMiddleware(limits, codec))
+	}
+	if cfg.CacheResults {
+		queryCacheMiddleware, err := NewResultsCacheMiddleware(log, cfg.ResultsCacheConfig, limits, codec, cacheExtractor)
```

Comment:
Just to confirm, the utilization of `limits` in multiple middlewares will implement that parallelism at _each_ middleware function (n^m total max parallelism where `n = parallelism factor` and `m = number of middleware utilizing doRequests`). This is only until we have something more sophisticated to control centralized limits, yes?

Note: Currently cortex uses `doRequests` in [multiple places](https://github.com/cortexproject/cortex/search?q=doRequests&unscoped_q=doRequests) so it's not like you're introducing this behavior, but it's something I wanted to confirm.

Reply:
Correct !